File: src/main/java/work/stdpi/pge/editor/render/EditorUI.java
package work.stdpi.pge.editor.render;

import net.minecraft.client.MinecraftClient;
import net.minecraft.client.gui.DrawContext;
import org.lwjgl.glfw.GLFW;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import work.stdpi.pge.editor.logic.EditorManager;
import work.stdpi.pge.editor.logic.ViewportController;
import work.stdpi.pge.editor.WindowAccessor;

public class EditorUI {
    public static final EditorUI INSTANCE = new EditorUI();
    private static final Logger LOGGER = LoggerFactory.getLogger("pge-editor/ui");
    private final TerminalRenderer term = new TerminalRenderer();
    private final GlyphGridRenderer glyphGrid = new GlyphGridRenderer();
    private final MiniGameRenderer miniGame = new MiniGameRenderer();
    private boolean regionLogged = false;
    private boolean focused = false;
    private boolean windowFocused = true;
    private int ex, ey, ew, eh;

    public void render(DrawContext context) {
        if (!EditorManager.INSTANCE.isEnabled()) return;
        var mc = MinecraftClient.getInstance();
        var win = mc.getWindow();
        
        int rsw = ((WindowAccessor) (Object) win).pge$getRealScaledWidth();
        int rsh = ((WindowAccessor) (Object) win).pge$getRealScaledHeight();
        int rfw = ((WindowAccessor) (Object) win).pge$getRealFramebufferWidth();
        int rfh = ((WindowAccessor) (Object) win).pge$getRealFramebufferHeight();

        // EDITOR zone
        int gx = ViewportController.INSTANCE.getX();
        int gy = ViewportController.INSTANCE.getY();
        int gw = ViewportController.INSTANCE.getWidth();
        int gh = ViewportController.INSTANCE.getHeight();

        switch (EditorManager.INSTANCE.getSide()) {
            case LEFT ->  { ex = 0; ey = 0; ew = gx; eh = rsh; }
            case RIGHT -> { ex = gx + gw; ey = 0; ew = rsw - ex; eh = rsh; }
            case TOP ->   { ex = 0; ey = 0; ew = rsw; eh = gy; }
            case BOTTOM ->{ ex = 0; ey = gy + gh; ew = rsw; eh = rsh - ey; }
        }

        if (ew <= 0 || eh <= 0) return;

        if (!regionLogged) {
            LOGGER.info(
                "editor region side={} editor=({}, {}) {}x{} viewport=({}, {}) {}x{} screen={}x{}",
                EditorManager.INSTANCE.getSide(),
                ex, ey, ew, eh,
                gx, gy, gw, gh,
                rsw, rsh
            );
            regionLogged = true;
        }

        boolean nowWindowFocused = GLFW.glfwGetWindowAttrib(win.getHandle(), GLFW.GLFW_FOCUSED) == GLFW.GLFW_TRUE;
        if (nowWindowFocused && !windowFocused && focused) {
            LOGGER.info("minecraft window focus restored while terminal focused");
            term.focus();
        }
        windowFocused = nowWindowFocused;

        if (EditorManager.INSTANCE.getRenderMode() == EditorManager.RenderMode.TERMINAL) {
            float framebufferScaleX = rsw > 0 ? (float) rfw / rsw : 1.0f;
            float framebufferScaleY = rsh > 0 ? (float) rfh / rsh : 1.0f;
            int exPx = Math.round(ex * framebufferScaleX);
            int eyPx = Math.round(ey * framebufferScaleY);
            int ewPx = Math.max(1, Math.round(ew * framebufferScaleX));
            int ehPx = Math.max(1, Math.round(eh * framebufferScaleY));

            context.getMatrices().pushMatrix();
            context.getMatrices().scale(1.0f / framebufferScaleX, 1.0f / framebufferScaleY);
            context.fill(exPx, eyPx, exPx + ewPx, eyPx + ehPx, 0xFF1E1E1E);
            term.render(context, exPx, eyPx, ewPx, ehPx);
            context.getMatrices().popMatrix();
        } else if (EditorManager.INSTANCE.getRenderMode() == EditorManager.RenderMode.GIZMO_GRID) {
            float liedScaleX = (float) win.getScaledWidth() / rsw;
            float liedScaleY = (float) win.getScaledHeight() / rsh;
            context.getMatrices().pushMatrix();
            context.getMatrices().scale(liedScaleX, liedScaleY);
            context.fill(ex, ey, ex + ew, ey + eh, 0xFF1E1E1E);
            glyphGrid.render(context, ex, ey, ew, eh);
            context.getMatrices().popMatrix();
        } else {
            float liedScaleX = (float) win.getScaledWidth() / rsw;
            float liedScaleY = (float) win.getScaledHeight() / rsh;
            context.getMatrices().pushMatrix();
            context.getMatrices().scale(liedScaleX, liedScaleY);
            context.fill(ex, ey, ex + ew, ey + eh, 0xFF1E1E1E);
            miniGame.render(context, ex, ey, ew, eh);
            context.getMatrices().popMatrix();
        }
    }

    public boolean onMouse(int b, int a, int m) {
        if (!EditorManager.INSTANCE.isEnabled()) return false;
        var mc = MinecraftClient.getInstance();
        var win = mc.getWindow();
        if (EditorManager.INSTANCE.getRenderMode() == EditorManager.RenderMode.TERMINAL) {
            float framebufferScaleX = ((WindowAccessor)(Object)win).pge$getRealScaledWidth() > 0
                ? (float) ((WindowAccessor)(Object)win).pge$getRealFramebufferWidth() / ((WindowAccessor)(Object)win).pge$getRealScaledWidth()
                : 1.0f;
            float framebufferScaleY = ((WindowAccessor)(Object)win).pge$getRealScaledHeight() > 0
                ? (float) ((WindowAccessor)(Object)win).pge$getRealFramebufferHeight() / ((WindowAccessor)(Object)win).pge$getRealScaledHeight()
                : 1.0f;
            int exPx = Math.round(ex * framebufferScaleX);
            int eyPx = Math.round(ey * framebufferScaleY);
            int ewPx = Math.max(1, Math.round(ew * framebufferScaleX));
            int ehPx = Math.max(1, Math.round(eh * framebufferScaleY));
            int mx = (int) mc.mouse.getX();
            int my = (int) mc.mouse.getY();

            if (mx >= exPx && mx < exPx + ewPx && my >= eyPx && my < eyPx + ehPx) {
                focused = true;
                LOGGER.info("terminal mouse focus acquired at {},{}", mx - exPx, my - eyPx);
                term.focus();
                term.onMouse(mx - exPx, my - eyPx, b, a, m, ewPx, ehPx);
                return true;
            }
            if (focused) {
                LOGGER.info("terminal mouse focus lost");
            }
            focused = false;
            return false;
        }
        double s = (double) ((WindowAccessor)(Object)win).pge$getRealFramebufferWidth() / ((WindowAccessor)(Object)win).pge$getRealScaledWidth();
        int mx = (int)(mc.mouse.getX() / s), my = (int)(mc.mouse.getY() / s);
        
        if (mx >= ex && mx < ex + ew && my >= ey && my < ey + eh) {
            focused = true;
            if (EditorManager.INSTANCE.getRenderMode() == EditorManager.RenderMode.TERMINAL) {
                LOGGER.info("terminal mouse focus acquired at {},{}", mx - ex, my - ey);
                term.focus();
                term.onMouse(mx - ex, my - ey, b, a, m, ew, eh);
            } else if (EditorManager.INSTANCE.getRenderMode() == EditorManager.RenderMode.GIZMO_GRID) {
                glyphGrid.onMouse(mx - ex, my - ey, b, a, m, ew, eh);
            } else {
                miniGame.onMouse(mx - ex, my - ey, b, a, m, ew, eh);
            }
            return true;
        }
        if (focused) {
            LOGGER.info("terminal mouse focus lost");
        }
        focused = false;
        return false;
    }

    public void onMove(double x, double y) {
        if (!EditorManager.INSTANCE.isEnabled()) return;
        var mc = MinecraftClient.getInstance();
        var win = mc.getWindow();
        if (EditorManager.INSTANCE.getRenderMode() == EditorManager.RenderMode.TERMINAL) {
            float framebufferScaleX = ((WindowAccessor)(Object)win).pge$getRealScaledWidth() > 0
                ? (float) ((WindowAccessor)(Object)win).pge$getRealFramebufferWidth() / ((WindowAccessor)(Object)win).pge$getRealScaledWidth()
                : 1.0f;
            float framebufferScaleY = ((WindowAccessor)(Object)win).pge$getRealScaledHeight() > 0
                ? (float) ((WindowAccessor)(Object)win).pge$getRealFramebufferHeight() / ((WindowAccessor)(Object)win).pge$getRealScaledHeight()
                : 1.0f;
            int exPx = Math.round(ex * framebufferScaleX);
            int eyPx = Math.round(ey * framebufferScaleY);
            int ewPx = Math.max(1, Math.round(ew * framebufferScaleX));
            int ehPx = Math.max(1, Math.round(eh * framebufferScaleY));
            int mx = (int) x;
            int my = (int) y;
            if (mx >= exPx && mx < exPx + ewPx && my >= eyPx && my < eyPx + ehPx) {
                term.onMove(mx - exPx, my - eyPx, ewPx, ehPx);
            }
            return;
        }
        double s = (double) ((WindowAccessor)(Object)win).pge$getRealFramebufferWidth() / ((WindowAccessor)(Object)win).pge$getRealScaledWidth();
        int mx = (int)(x / s);
        int my = (int)(y / s);
        if (mx >= ex && mx < ex + ew && my >= ey && my < ey + eh) {
            if (EditorManager.INSTANCE.getRenderMode() == EditorManager.RenderMode.TERMINAL) {
                term.onMove(mx - ex, my - ey, ew, eh);
            } else if (EditorManager.INSTANCE.getRenderMode() == EditorManager.RenderMode.GIZMO_GRID) {
                glyphGrid.onMove(mx - ex, my - ey, ew, eh);
            } else {
                miniGame.onMove(mx - ex, my - ey, ew, eh);
            }
        }
    }

    public boolean onKey(int k, int a, int m) {
        if (!EditorManager.INSTANCE.isEnabled()) return false;
        if (k == GLFW.GLFW_KEY_BACKSLASH) return false;
        boolean terminalGameplayCapture =
            EditorManager.INSTANCE.getRenderMode() == EditorManager.RenderMode.TERMINAL
                && MinecraftClient.getInstance().currentScreen == null;
        if (focused || terminalGameplayCapture) {
            if (EditorManager.INSTANCE.getRenderMode() == EditorManager.RenderMode.TERMINAL) {
                LOGGER.info("terminal key key={} action={} mods={}", k, a, m);
                term.onKey(k, a, m);
                return true;
            } else if (EditorManager.INSTANCE.getRenderMode() == EditorManager.RenderMode.GIZMO_GRID) {
                return glyphGrid.onKey(k, a);
            } else {
                return miniGame.onKey(k, a);
            }
        }
        return false;
    }

    public boolean onChar(int c) {
        boolean terminalGameplayCapture =
            EditorManager.INSTANCE.isEnabled()
                && EditorManager.INSTANCE.getRenderMode() == EditorManager.RenderMode.TERMINAL
                && MinecraftClient.getInstance().currentScreen == null;
        if (EditorManager.INSTANCE.isEnabled()
            && EditorManager.INSTANCE.getRenderMode() == EditorManager.RenderMode.TERMINAL
            && (focused || terminalGameplayCapture)) {
            LOGGER.info("terminal char codepoint={}", c);
            term.onChar(c);
            return true;
        }
        return false;
    }
}

File: src/main/kotlin/work/stdpi/pge/editor/render/EditorUI.kt
package work.stdpi.pge.editor.render

import net.minecraft.client.MinecraftClient
import net.minecraft.client.gui.DrawContext
import net.minecraft.client.util.Window
import org.lwjgl.glfw.GLFW
import work.stdpi.pge.editor.realFramebufferHeight
import work.stdpi.pge.editor.realFramebufferWidth
import work.stdpi.pge.editor.realMetrics
import work.stdpi.pge.editor.realScaledHeight
import work.stdpi.pge.editor.realScaledWidth
import work.stdpi.pge.editor.logic.EditorManager
import work.stdpi.pge.editor.logic.EditorManager.DockSide
import work.stdpi.pge.editor.logic.ViewportController
import work.stdpi.pge.editor.render.applet.AbstractTerminalApplet
import work.stdpi.pge.editor.render.canvas.CanvasRect
import work.stdpi.pge.editor.render.canvas.CanvasSpace
import work.stdpi.pge.editor.render.canvas.input.CanvasCharEvent
import work.stdpi.pge.editor.render.canvas.input.CanvasKeyEvent
import work.stdpi.pge.editor.render.canvas.input.CanvasPointerEvent
import work.stdpi.pge.editor.render.canvas.input.CanvasScrollEvent
import kotlin.math.max

object EditorUI {
  private var focused = false
  private var windowFocused = true
  private var editorBounds = CanvasRect.ZERO
  private var appletBounds = CanvasRect.ZERO

  fun render(context: DrawContext) {
    if (!EditorManager.isEnabled) return
    val mc = MinecraftClient.getInstance()
    val win = mc.window
    val metrics = WindowMetrics.from(win)
    val applet = EditorManager.currentApplet
    editorBounds = computeEditorBounds(metrics)
    if (!editorBounds.isVisible) return

    val nowWindowFocused =
        GLFW.glfwGetWindowAttrib(win.handle, GLFW.GLFW_FOCUSED) == GLFW.GLFW_TRUE
    if (nowWindowFocused && !windowFocused && focused) {
      applet.onCanvasFocusGained()
    } else if (!nowWindowFocused && windowFocused) {
      if (focused) applet.onCanvasFocusLost()
      focused = false
    }
    windowFocused = nowWindowFocused

    appletBounds =
        when (applet.space) {
          CanvasSpace.FRAMEBUFFER -> editorBounds.toFramebuffer(metrics)
          CanvasSpace.SCALED -> editorBounds
        }

    context.matrices.pushMatrix()
    when (applet.space) {
      CanvasSpace.FRAMEBUFFER ->
          context.matrices.scale(1.0f / metrics.framebufferScaleX, 1.0f / metrics.framebufferScaleY)
      CanvasSpace.SCALED -> context.matrices.scale(metrics.liedScaleX, metrics.liedScaleY)
    }
    context.fill(
        appletBounds.x,
        appletBounds.y,
        appletBounds.x + appletBounds.width,
        appletBounds.y + appletBounds.height,
        PANEL_BG)
    applet.render(context, appletBounds)
    context.matrices.popMatrix()
  }

  fun onMouse(b: Int, a: Int, m: Int): Boolean {
    if (!EditorManager.isEnabled) return false
    val mc = MinecraftClient.getInstance()
    val applet = EditorManager.currentApplet
    val (mx, my) =
        when (applet.space) {
          CanvasSpace.FRAMEBUFFER -> mc.mouse.x.toInt() to mc.mouse.y.toInt()
          CanvasSpace.SCALED -> WindowMetrics.from(mc.window).framebufferToScaled(mc.mouse.x, mc.mouse.y)
        }

    if (appletBounds.contains(mx, my)) {
      if (!focused) {
        applet.onCanvasFocusGained()
      }
      focused = true
      return applet.onCanvasPointerButton(
          CanvasPointerEvent(
              localX = mx - appletBounds.x,
              localY = my - appletBounds.y,
              bounds = appletBounds,
              button = b,
              action = a,
              mods = m))
    }
    if (focused) applet.onCanvasFocusLost()
    focused = false
    return false
  }

  fun shouldBlockGameMouseInput(): Boolean {
    if (!EditorManager.isEnabled) return false
    if (MinecraftClient.getInstance().currentScreen != null) return false
    return focused
  }

  fun shouldBlockGameKeyboardInput(key: Int? = null): Boolean {
    if (!EditorManager.isEnabled) return false
    if (key != null && EditorManager.isToggleKey(key)) return false
    return focused
  }

  fun onMove(x: Double, y: Double) {
    if (!EditorManager.isEnabled || !focused) return
    val applet = EditorManager.currentApplet
    val (mx, my) =
        when (applet.space) {
          CanvasSpace.FRAMEBUFFER -> x.toInt() to y.toInt()
          CanvasSpace.SCALED ->
              WindowMetrics.from(MinecraftClient.getInstance().window).framebufferToScaled(x, y)
        }
    if (appletBounds.contains(mx, my)) {
      applet.onCanvasPointerMove(
          CanvasPointerEvent(
              localX = mx - appletBounds.x,
              localY = my - appletBounds.y,
              bounds = appletBounds))
    }
  }

  fun onScroll(horizontalAmount: Double, verticalAmount: Double): Boolean {
    if (!EditorManager.isEnabled || !focused) return false
    val mc = MinecraftClient.getInstance()
    val applet = EditorManager.currentApplet
    val (mx, my) =
        when (applet.space) {
          CanvasSpace.FRAMEBUFFER -> mc.mouse.x.toInt() to mc.mouse.y.toInt()
          CanvasSpace.SCALED -> WindowMetrics.from(mc.window).framebufferToScaled(mc.mouse.x, mc.mouse.y)
        }

    if (appletBounds.contains(mx, my)) {
      return applet.onCanvasScroll(
          CanvasScrollEvent(
              localX = mx - appletBounds.x,
              localY = my - appletBounds.y,
              bounds = appletBounds,
              horizontalAmount = horizontalAmount,
              verticalAmount = verticalAmount,
              mods = 0))
    }
    return false
  }

  fun onKey(k: Int, a: Int, m: Int): Boolean {
    if (!EditorManager.isEnabled) return false
    if (EditorManager.isToggleKey(k)) return false
    val applet = EditorManager.currentApplet
    if (focused) {
      return applet.onCanvasKey(CanvasKeyEvent(k, a, m))
    }
    return false
  }

  fun onChar(c: Int): Boolean {
    val applet = EditorManager.currentApplet
    if (EditorManager.isEnabled && focused) {
      return applet.onCanvasChar(CanvasCharEvent(c))
    }
    return false
  }

  private fun computeEditorBounds(metrics: WindowMetrics): CanvasRect {
    val gx = ViewportController.x
    val gy = ViewportController.y
    val gw = ViewportController.width
    val gh = ViewportController.height
    return when (EditorManager.side) {
      DockSide.LEFT -> CanvasRect(0, 0, gx, metrics.scaledHeight)
      DockSide.RIGHT -> CanvasRect(gx + gw, 0, metrics.scaledWidth - (gx + gw), metrics.scaledHeight)
      DockSide.TOP -> CanvasRect(0, 0, metrics.scaledWidth, gy)
      DockSide.BOTTOM -> CanvasRect(0, gy + gh, metrics.scaledWidth, metrics.scaledHeight - (gy + gh))
    }
  }

  private fun CanvasRect.toFramebuffer(metrics: WindowMetrics): CanvasRect =
      CanvasRect(
          Math.round(x * metrics.framebufferScaleX),
          Math.round(y * metrics.framebufferScaleY),
          max(1, Math.round((x + width) * metrics.framebufferScaleX) - Math.round(x * metrics.framebufferScaleX)),
          max(1, Math.round((y + height) * metrics.framebufferScaleY) - Math.round(y * metrics.framebufferScaleY)))

  private data class WindowMetrics(
      val scaledWidth: Int,
      val scaledHeight: Int,
      val framebufferScaleX: Float,
      val framebufferScaleY: Float,
      val liedScaleX: Float,
      val liedScaleY: Float
  ) {
    fun framebufferToScaled(x: Double, y: Double): Pair<Int, Int> {
      val scaledX = if (framebufferScaleX != 0.0f) x / framebufferScaleX else x
      val scaledY = if (framebufferScaleY != 0.0f) y / framebufferScaleY else y
      return scaledX.toInt() to scaledY.toInt()
    }

    companion object {
      fun from(window: Window): WindowMetrics {
        val real = window.realMetrics()
        val scaledWidth = real.realScaledWidth
        val scaledHeight = real.realScaledHeight
        val framebufferWidth = real.realFramebufferWidth
        val framebufferHeight = real.realFramebufferHeight
        val framebufferScaleX = if (scaledWidth > 0) framebufferWidth.toFloat() / scaledWidth else 1.0f
        val framebufferScaleY =
            if (scaledHeight > 0) framebufferHeight.toFloat() / scaledHeight else 1.0f
        return WindowMetrics(
            scaledWidth = scaledWidth,
            scaledHeight = scaledHeight,
            framebufferScaleX = framebufferScaleX,
            framebufferScaleY = framebufferScaleY,
            liedScaleX = if (scaledWidth > 0) window.scaledWidth.toFloat() / scaledWidth else 1.0f,
            liedScaleY = if (scaledHeight > 0) window.scaledHeight.toFloat() / scaledHeight else 1.0f)
      }
    }
  }
}

private const val PANEL_BG = -0xe1e1e2
